File: backend/src/main/java/com/codeloom/backend/service/TopicService.java
package com.codeloom.backend.service;

import com.codeloom.backend.dao.topic.TopicRepository;
import com.codeloom.backend.dao.topic.TopicRepositoryCustomImpl;
import com.codeloom.backend.dto.CreateTopicRequest;
import com.codeloom.backend.dto.UpdateProblemTopicRequest;
import com.codeloom.backend.dto.UpdateTopicRequest;
import com.codeloom.backend.model.ProblemTopicRelationship;
import com.codeloom.backend.model.Topic;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

@RequiredArgsConstructor
@Service
public class TopicService {
    private final TopicRepositoryCustomImpl topicRepositoryCustom;
    private final TopicRepository topicRepository;

    public Iterable<Topic> findAll() {
        return topicRepository.findAll();
    }

    public Topic findById(UUID topicId) {
        return topicRepository
                .findById(topicId)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND, "Entity with id `" + topicId + "` not found"));
    }

    @Transactional
    public Topic create(CreateTopicRequest request) {
        return topicRepository.save(Topic.builder().name(request.name()).build());
    }

    @Transactional
    public void createManyWithProblem(long problemId, List<UpdateProblemTopicRequest> topics) {
        topicRepositoryCustom.deleteRelationshipsWithProblem(problemId);
        List<ProblemTopicRelationship> relationships = new ArrayList<>();

        for (UpdateProblemTopicRequest topic : topics) {
            UUID topicId = null;
            if (topic.topic_id() != null) {
                try {
                    topicId = UUID.fromString(topic.topic_id());
                } catch (IllegalArgumentException ignored) {
                    // Ignore malformed references, matching missing topic references.
                }
            } else if (topic.name() != null) {
                topicId = topicRepository
                        .save(Topic.builder().name(topic.name()).build())
                        .getId();
            }

            if (topicId != null) {
                relationships.add(ProblemTopicRelationship.builder()
                        .topicId(topicId)
                        .problemId(problemId)
                        .build());
            }
        }
        topicRepositoryCustom.saveAllProblemRelationships(relationships);
    }

    @Transactional
    public Topic update(UUID topicId, UpdateTopicRequest request) {
        Topic topic = findById(topicId).withName(request.name());
        return topicRepository.save(topic);
    }

    @Transactional
    public void delete(UUID topicId) {
        topicRepository.deleteById(topicId);
    }
}

File: backend/src/main/kotlin/com/codeloom/backend/service/TopicService.kt
package com.codeloom.backend.service

import com.codeloom.backend.dao.topic.TopicRepository
import com.codeloom.backend.dao.topic.TopicRepositoryCustomImpl
import com.codeloom.backend.dto.CreateTopicRequest
import com.codeloom.backend.model.ProblemTopicRelationship
import com.codeloom.backend.model.Topic
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.server.ResponseStatusException
import tools.jackson.databind.JsonNode
import tools.jackson.databind.ObjectMapper
import java.util.*

@Service
class TopicService(
    private val topicRepositoryCustomImpl: TopicRepositoryCustomImpl,
    private val topicRepository: TopicRepository,
    private val objectMapper: ObjectMapper,
) {
    fun findAll(): Iterable<Topic> = topicRepository.findAll()

    fun findById(id: UUID): Topic = findOrThrow(id)

    @Transactional
    fun create(request: CreateTopicRequest): Topic {
        val topic = Topic(name = request.name)
        return topicRepository.save(topic)
    }

    @Transactional
    fun createManyWithProblem(
        problemId: Long,
        node: JsonNode,
    ) {
        topicRepositoryCustomImpl.deleteRelationshipsWithProblem(problemId)
        val problemTopicRelationships =
            node.asIterable()
                .mapNotNull {
                    when {
                        it.has("topic_id") -> {
                            try {
                                UUID.fromString(it["topic_id"].asString())
                            } catch (_: IllegalArgumentException) {
                                null
                            }
                        }

                        it.has("name") -> {
                            val topic = Topic(name = it["name"].asString())
                            topicRepository.save(topic).id!!
                        }

                        else -> null
                    }
                }
                .map { ProblemTopicRelationship(it, problemId) }
        topicRepositoryCustomImpl.saveAllProblemRelationships(problemTopicRelationships)
    }

    @Transactional
    fun patch(
        id: UUID,
        patchNode: JsonNode,
    ): Topic {
        val topic = findOrThrow(id)
        objectMapper.readerForUpdating(topic).readValue<Topic>(patchNode)
        return topicRepository.save(topic)
    }

    @Transactional
    fun delete(id: UUID) {
        topicRepository.deleteById(id)
    }

    private fun findOrThrow(id: UUID): Topic {
        return topicRepository.findById(id)
            .orElseThrow {
                ResponseStatusException(HttpStatus.NOT_FOUND, "Entity with id `$id` not found")
            }
    }
}
